### ISSUE TITLE
Pageable results and @Query annotation 

### ISSUE THREAD

**Issue by edovac** [spring-projects/spring-data-elasticsearch#1843]
Spring Boot 2.4.6 and 2.4.7 broke compatibility between Page<T> result type and @Query annotation on same method.

A method with this signature used to work up to Spring Boot 2.4.5:

    @Query("{\"match_phrase_prefix\":{\"stringValue\":\"?0\"}}")
    Page<MyDomainObj> findByStringPrefix(String stringValue, Pageable pageable);

Starting from Spring Boot 2.4.6 whenever the code access to the objects inside the List a class cast exception is thrown.

See the attached test project.
The problem is shown with current setup. To replace it's behaviour with the previous version just replace Spring Boot version in pom.xml

[demo.zip](https://github.com/spring-projects/spring-data-elasticsearch/files/6649394/demo.zip)

**Comment by sothawo** [comment]
This regression seems to be caused by the backport of #1811 into the 4.1 branch where the processing of `@Query` annotataed methods returns now a `SearchPage<T>` which basically is a `Page<SearchHit<T>>`.

That this in version 4.2 works without an error comes from a previous change in 4.2 that cleaned up old code. I will add a fix to the 4.1 branch.

What you could do with the current version: Change the return type of your repository method to:

```java
@Query("{\"match_phrase_prefix\":{\"stringValue\":\"?0\"}}")
SearchPage<MyDomainObj> findByStringPrefix(String stringValue, Pageable pageable);
```
 and then in your code work with the `SearchHit<T>` objects; you could get a list of the entites with:

```java
SearchPage<MyDomainObj> result = repository.findByStringPrefix("a ", Pageable.unpaged());
List<MyDomainObj> list = result.getContent().stream().map(SearchHit::getContent).collect(Collectors.toList());
```

**Comment by edovac** [comment]
As I understand the `SearchPage<T>` will be the new standard for version 4.2.
The old convention seemed more appropriate because it allowed returning `Page<T>` directly from REST methods thus integrating REST layer and Spring Data layer. Maybe I'm missing something in the latest releases.

We have only a few points where whe use the `@Query` annotation with `Pageable` and I think we can refactor the code.

Thanks

**Comment by sothawo** [comment]
If you define your method to return a `SearchPage<T>`you get a `SearchPage<T>` which is a `Page<SearchHit<T>>`. If the return type is `Page<T>` you get a `Page`. That this produces an error in your case was a regression as I wrote.
This is fixed in the 4.1.x branch and will be contained in the next SR

**Comment by edovac** [comment]
Ok, this is clear. Thanks
